### PR TITLE
Task 3554, enhance xcatprobe discovery precheck

### PR DESCRIPTION
**Pull request by xuweibj**
#3554 

1. Add option ``-c`` to do pre-check only.
2. Check attribute ``chain`` whether null.
```
# xcatprobe discovery -n test -m switch -c
[test_node1] : Invalid value for attribute 'chain'                                                                [FAIL]
```
3. Check genesis version whether built in fedora 26.
```
# xcatprobe discovery -n test_node -m mtms -c
Genesis files are avaliable                                                                                       [ OK ]
xcat-genesis-base-ppc64 is built in environment fc26 on ppc64.                                                    [WARN]
```
4. Check genesis files when NIC is "noboot".
```
"dhcpinterfaces","enp0s1,enp0s2:noboot",,
# ip -4 -o a | grep enp0s2
3: enp0s2    inet 50.3.5.4/8 brd 50.255.255.255 scope global enp0s2\       valid_lft forever preferred_lft forever

# xcatprobe discovery -n test_node -m mtms -c
Genesis files are avaliable                                                                                       [ OK ]
File /tftpboot/pxelinux.cfg/p/50.0.0.0_8 should not exist, please delete it.                                      [WARN]
File /tftpboot/xcat/xnba/nets/50.0.0.0_8 should not exist, please delete it.                                      [WARN]
File /tftpboot/xcat/xnba/nets/50.0.0.0_8.elilo should not exist, please delete it.                                [WARN]
```
5. Check switch & port whether valid.
```
# xcatprobe discovery -n test -m switch -c
[test_node1] : Attribute switchport Invalid                                                                       [FAIL]
[test_node2] : No switch frame31sw1 information obtained.                                                         [FAIL]
```
6. Check IP parse.
```
# xcatprobe discovery  -n test -m mtms -c
DHCP dynamic range is configured                                                                                  [FAIL]
Unknown host test_node2, please run 'makehosts' and 'makedns'
```
7. Check after genesis update, whether run mknb.
```
# xcatprobe discovery -n test -m mtms -c
Genesis files are avaliable                                                                                       [ OK ]
Genesis packages have been updated, please run 'mknb <arch>'                                                      [WARN]
```

### CODE REVIEW

File: xCAT-probe/subcmds/discovery [hunk 1]
```diff
@@ -194,17 +208,27 @@ sub check_pre_defined_node {
...
+        }
     }
     if (@bmcnodes) {
         my $bmcnode = join(",", @bmcnodes);
         probe_utils->send_msg("stdout", "f", "[$bmcnode] : bmc node(s)");
         $rst = 1; 
     }
 
+    if (@nochain) {
+        my $nochainnode = join(",", @nochain);
```
Comment (hu-weihua) [inline xCAT-probe/subcmds/discovery]: Suggest to add below line after line 277 to remove the last ``,``.
   
     $nochainnode =~ s/,$//g;

File: xCAT-probe/subcmds/discovery [hunk 2]
```diff
@@ -322,8 +350,13 @@ sub check_pre_defined_node {
...
                 if (!(exists($nodecheckrst{$node}{"switch"})) or $nodecheckrst{$node}{"switch"} !~ /^\w/) {
                     push @error_attribute, "switch";
+                } elsif ($error_switch =~ $nodecheckrst{$node}{"switch"}) {
+                    my $error_switch_string = "No switch " . $nodecheckrst{$node}{"switch"} . " information obtained."; 
+                    push @{$error_switch_hash{$error_switch_string}}, $node;
+                    next;
                 }
-                if (!(exists($nodecheckrst{$node}{"switchport"})) or $nodecheckrst{$node}{"switchport"} !~ /^\w/) {
+
+                if (!(exists($nodecheckrst{$node}{"switchport"})) or $nodecheckrst{$node}{"switchport"} !~ /^\w/ or $valid_nodes !~ $node) {
```
Comment (hu-weihua) [inline xCAT-probe/subcmds/discovery]: What is we want to check? The predefined node has wrong ``switch/switchport`` configuration?  If I configured ``node1  switch=switch1 switchport=1``, but in fact, ``node1`` is plugged in ``switch1``'s port 2. Can line 359  find this problem?

File: xCAT-probe/subcmds/discovery [hunk 3]
```diff
@@ -322,8 +350,13 @@ sub check_pre_defined_node {
 
                 if (!(exists($nodecheckrst{$node}{"switch"})) or $nodecheckrst{$node}{"switch"} !~ /^\w/) {
                     push @error_attribute, "switch";
+                } elsif ($error_switch =~ $nodecheckrst{$node}{"switch"}) {
+                    my $error_switch_string = "No switch " . $nodecheckrst{$node}{"switch"} . " information obtained.";
```
Comment (hu-weihua) [inline xCAT-probe/subcmds/discovery]: Maybe from customer point of view. The error message should be 
   
   my $error_switch_string = "Switch " . $nodecheckrst{$node}{"switch"} . " can not be accessible.";

File: xCAT-probe/subcmds/discovery [hunk 4]
```diff
@@ -377,6 +410,11 @@ sub check_pre_defined_node {
             probe_utils->send_msg("stdout", "f", "[$switchnode] : Duplicate node definition found for the same switch and switchport $switch_port.");
             $rst = 1;
         }
+        foreach my $key (keys %error_switch_hash) {
+            my $nodes = join(",", @{$error_switch_hash{$key}});
```
Comment (hu-weihua) [inline xCAT-probe/subcmds/discovery]: Suggest to remove the last ``,``.

File: xCAT-probe/subcmds/discovery [hunk 5]
```diff
@@ -518,6 +559,40 @@ sub check_genesis_file {
...
+        $genesis_time = $genesis_script_time;
+    }
+
+    my $dhcpinterface = `tabdump site | sed s/'"'//g | grep '^dhcpinterfaces'`;
+    my $noboot_file;
+    my @tmp_info = split(",", $dhcpinterface);
+    foreach (@tmp_info) {
+        if ($_ =~ /(.+):noboot/) {
+            my $noboot_nic = $1; 
+            my $ip_string = `ip -4 -o a | grep $noboot_nic | awk -F ' ' '{print \$4}'`;
```
Comment (hu-weihua) [inline xCAT-probe/subcmds/discovery]: This line can be refined to 

     ip -4 -o a |awk -F' ' '/$noboot_nic/ {print $4}'

File: xCAT-probe/subcmds/discovery [hunk 6]
```diff
@@ -518,6 +559,40 @@ sub check_genesis_file {
...
+            if ($ip_string) {
+                my ($noboot_ip, $noboot_netmask) = split('\/', $ip_string);
+                my @noboot_ips = split('\.', $noboot_ip);
+                $noboot_file = $noboot_ips[0] . ".0.0.0_$noboot_netmask";
```
Comment (hu-weihua) [inline xCAT-probe/subcmds/discovery]: I am not sure this logic which used to calculate the ``noboot_file`` is correct. I think the network should be calculated by ip and netmask, not just be combined simply by ``$noboot_ips[0] . ".0.0.0_$noboot_netmask``

File: xCAT-probe/subcmds/discovery [hunk 7]
```diff
@@ -518,6 +559,40 @@ sub check_genesis_file {
...
+        push @warn_msg, "xcat-genesis-base-ppc64 is built in environment fc26 on $1.";
+    }
+
+    my $genesis_update_flag_p;
+    my $genesis_update_flag_x;
+    my $genesis_time;
+    my $genesis_base_time;
+    my $genesis_script_time;
+    $genesis_base_time  = `stat /etc/xcat/genesis-base-updated | grep Modify | cut -d ' ' -f 2-3` if (-e '/etc/xcat/genesis-base-updated');
+    $genesis_script_time = `stat /etc/xcat/genesis-scripts-updated | grep Modify | cut -d ' ' -f 2-3` if (-e '/etc/xcat/genesis-scripts-updated');
```
Comment (hu-weihua) [inline xCAT-probe/subcmds/discovery]: What the progress would be if these 2 files ``/etc/xcat/genesis-base-updated`` and ``/etc/xcat/genesis-scripts-updated`` does not exist.  ``$genesis_base_time`` and ``$genesis_script_time`` won't have value.

File: xCAT-probe/subcmds/discovery [hunk 8]
```diff
@@ -614,8 +708,8 @@ sub check_genesis_file {
                     next;
                 }
 
-                unless (open(FILE, $_)) {
-                    push @errors, "Cannot open file $_.";
+                unless (open(FILE, $file)) {
+                    push @errors, "Cannot open file $file.";
```
Comment (hu-weihua) [inline xCAT-probe/subcmds/discovery]: It is better to change line 712  to 

    push @errors, "Cannot open file $file : $!.";

File: xCAT-probe/subcmds/discovery (hunk 2 above)
Comment (xuweibj) [inline xCAT-probe/subcmds/discovery]: Only can check the node's switchport does not exist.

File: xCAT-probe/subcmds/discovery (hunk 7 above)
Comment (xuweibj) [inline xCAT-probe/subcmds/discovery]: Added judgement on line 650&775.

**Comment by zet809** [comment]
Looks good to me, I agree to merge.

**Pull request merged**